KAFKA-9830: Implement AutoCloseable in ErrorReporter and subclasses

gharris1727 opened the pull request:
* The DeadLetterQueueReporter has a KafkaProducer that it must close to clean up resources
* Currently, the producer and its threads are leaked every time a task is stopped
* Responsibility for cleaning up ErrorReporters is transitively assigned to the
    ProcessingContext, RetryWithToleranceOperator, and WorkerSinkTask/WorkerSinkTask classes
* One new unit test in ErrorReporterTest asserts that the producer is closed by the dlq reporter

Signed-off-by: Greg Harris <gregh@confluent.io>

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

gharris1727 committed:
KAFKA-9830: Implement AutoCloseable in ErrorReporter and subclasses

* The DeadLetterQueueReporter has a KafkaProducer that it must close to clean up resources
* Currently, the producer and it's threads are leaked every time a task is stopped
* Responsibility for cleaning up ErrorReporters is transitively assigned to the
    ProcessingContext, RetryWithToleranceOperator, and WorkerSinkTask/WorkerSinkTask classes
* One new unit test in ErrorReporterTest asserts that the producer is closed by the dlq reporter

Signed-off-by: Greg Harris <gregh@confluent.io>

gharris1727 commented:
cc @wicknicks @C0urante @ncliang for review

wicknicks commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/LogReporter.java:
```diff
@@ -69,4 +69,7 @@ String message(ProcessingContext context) {
     }
 
+    @Override
+    public void close() {
```
want to add this as a default implementation in the ErrorReporter interface?

C0urante replied:
It might save us a few LOC if we added a default implementation but I kind of like the guarantee that any new implementation of the `ErrorReporter` interface would have to explicitly add a `close` method since it might help prevent resource leaks similar to the ones that warranted these changes in the future. And it's not public API, so we don't have to worry about breaking compatibility by adding a new method with no default implementation.

gharris1727 replied:
I'll revert this to reduce the number of files changed and preserve backwards compatibility, but I would be fine with either strategy.

When implementing this method for the LogReporter, I did for a moment have to scan the file and consider what resources might need to be cleaned up. I appreciate that the non-default method forced me to do that, and when/if another ErrorReporter subclass is written, I think it will be helpful too.

wicknicks reviewed (commented):
thanks for the quick fix, @gharris1727. one nit.

gharris1727 committed:
Make no-op close a default method in ErrorReporter for backwards-compatibility

Signed-off-by: Greg Harris <gregh@confluent.io>

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/errors/ErrorReporterTest.java:
```diff
@@ -146,6 +146,20 @@ public void testReportDLQTwice() {
     }
 
+    @Test
+    public void testCloseDLQ() {
```
If I read this correctly, the guarantee here is that if a DLQ reporter is closed, its internal Kafka producer will also be closed.

Do you think it would also make sense to add a test to ensure that when a `WorkerSinkTask` and `WorkerSourceTask`  are shut down, and/or a `RetryWithToleranceOperator` is closed, the underlying `ErrorReporter` instances are closed as well?

gharris1727 replied:
I found ErrorHandlingTaskTest that seemed to be doing similar testing, and I managed to test the rest of the hierarchy (WorkerSourceTask/WorkerSinkTask -> RetryWithToleranceOperator -> ProcessingContext -> ErrorReporter) in just two tests.

Combined with the DLQ reporter -> KafkaProducer test from the first commit, I think the testing is adequate now.

C0urante replied:
Agreed, looks great!

gharris1727 committed:
Add test that WorkerSourceTask/WorkerSinkTask close ErrorReporters on task stop.

Signed-off-by: Greg Harris <gregh@confluent.io>

gharris1727 commented:
Thanks for your comments @wicknicks @C0urante, this is ready for a second pass.

C0urante reviewed (approved):
Still have a slight preference against adding a default implementation for `ErrorReporter::close` but will leave that to greater minds. LGTM, nice find @gharris1727!

wicknicks reviewed (approved):
Thanks! LGTM!

kkonstantine commented:
ok to test

chia7712 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ProcessingContext.java:
```diff
@@ -216,4 +216,10 @@ public void reporters(Collection<ErrorReporter> reporters) {
+    @Override
+    public void close() throws Exception {
+        for (ErrorReporter reporter : reporters) {
+            reporter.close();
```
The exception caused by reporter will break this loop. Should we keep the exception and then throw it after all reporters are closed?

gharris1727 replied:
Yes, thanks for catching this oversight. I've added logic to suppress the exceptions from each of the reporters, but i'm not sure if that's the idiomatic way to combine multiple exceptions. I've also added a test to verify that this error collection behavior is present.

chia7712 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ProcessingContext.java:
```diff
@@ -216,4 +216,10 @@ public void reporters(Collection<ErrorReporter> reporters) {
     }
 
+    @Override
+    public void close() throws Exception {
```
Is ```Exception``` in method signature necessary? the method signature of ```ErrorReporter#close``` don't include checked exception.

gharris1727 replied:
You're right, i've removed the throws annotation here and above. Thanks!

gharris1727 committed:
Remove throws from new close methods where no checked exception is thrown.

Signed-off-by: Greg Harris <gregh@confluent.io>

gharris1727 committed:
Handle exceptions from closing each reporter to collect as suppressed exceptions

Also add a test to verify that multiple reporters are closed despite exceptions

Signed-off-by: Greg Harris <gregh@confluent.io>

gharris1727 commented:
@kkonstantine It appears that the tests failed because of some unrelated checkstyle issue on trunk that has been fixed, can you re-run the tests?

kkonstantine commented:
ok to test

kkonstantine commented:
retest this please

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java:
```diff
@@ -170,6 +170,11 @@ protected void close() {
+            retryWithToleranceOperator.close();
+        } catch (Throwable t) {
+            log.warn("Could not close retry operator", t);
+        }
```
This is common enough that there's a util for that and is used extensively:

```suggestion
        Utils.closeQuietly(retryWithToleranceOperator, "retry operator");
```

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -175,6 +175,11 @@ protected void close() {
+            retryWithToleranceOperator.close();
+        } catch (Throwable t) {
+            log.warn("Could not close retry operator", t);
+        }
```
```suggestion
        Utils.closeQuietly(retryWithToleranceOperator, "retry operator");        
```

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ProcessingContext.java:
```diff
@@ -216,4 +217,18 @@ public void reporters(Collection<ErrorReporter> reporters) {
 
+    @Override
+    public void close() {
+        ConnectException e = new ConnectException("Failed to close all reporters");
+        for (ErrorReporter reporter : reporters) {
+            try {
+                reporter.close();
+            } catch (Throwable t) {
+                e.addSuppressed(t);
+            }
+        }
+        if (e.getSuppressed().length > 0) {
+            throw e;
+        }
```
Regardless how high performance we want this code to be, as a matter of principle we shouldn't create an exception (and pay the cost of filling its stacktrace during instantiation) if we don't need to throw it.

I sketched a modification just with a +1 line, but feel free to adjust to your style of preference.

```suggestion
        ConnectException e = null;
        for (ErrorReporter reporter : reporters) {
            try {
                reporter.close();
            } catch (Throwable t) {
                e = e == null ? new ConnectException("Failed to close all reporters") : e;
                e.addSuppressed(t);
            }
        }
        if (e != null) {
            throw e;
        }
```

kkonstantine replied:
Here's a nice blog re: exception costs and when they occur: 
https://shipilev.net/blog/2014/exceptional-performance

kkonstantine reviewed (commented):
Thanks @gharris1727 !
Important not to leak producers when tasks close

I have a couple suggestions to make this more concise and a recommendation to avoid instantiating an exception if it's not going to be used.

gharris1727 committed:
Use Utils.closeQuietly and defer creation of exception in loop

Signed-off-by: Greg Harris <gregh@confluent.io>

gharris1727 commented:
@kkonstantine I've implemented your feedback and this is ready for another pass.

kkonstantine commented:
retest this please

kkonstantine reviewed (approved):
jdk14 green, jdk8 2 failures in streams tests, jdk11 4 failures, also unrelated. 

LGTM!
